### DISCUSSION TITLE
JsonRPC method Re-run

### DISCUSSION THREAD

**Discussion by Garulf** [Flow-Launcher/Flow.Launcher#635] [Ideas]
With this method a plugin could tell Flow Launcher to rerun at a desired interval as long as the result has the rerun method. Similar to [Alfred's rerun functionality](https://www.alfredapp.com/blog/releases/alfred-3-2-released-a-whole-heap-of-improvements/#:~:text=Script%20filters%20have%20also%20seen%20significant%20improvements%3B%20Your%20script%20output%20can%20now%20tell%20Alfred%20to%20re-run%20the%20script%20after%20a%20set%20period%20of%20time.%20Take%20a%20look%20at%20the%20%22Advanced%20Script%20Filters%22%20workflow%20in%20Getting%20Started%20to%20see%20this%20in%20action.)

It can help with queries that change often and calls that are slow or have partial results.

**Comment by taooceros** [comment]
Interesting. So it is supposed to be background task?

**Comment by Garulf** [comment]
> Interesting. So it is supposed to be background task?

Basically has The launcher sleep for an interval provided by the plugin. Than reruns with the same query. Displaying any results on each rerun. 

```
While true:
    # display results
    if 'rerun' in results:
        sleep results['rerun']
    else:
        break
```

**Reply by taooceros** [reply]
Probably sleep feels weird because it forbid other plugin to run. But scheduling a rerun and wake up launcher is possible.

**Reply by Garulf** [reply]
Yeah sorry I was on mobile and just wanted to get out the general idea. 

Definitely would need to work in a non-blocking way. The user should be able to select what results have appeared so far or close flow launcher entirely stopping the reruns.

**Comment by pc223** [comment]
Interesting, what are the example use cases of this? 💭

**Reply by Garulf** [reply]
A common use case is for streaming data, like for a download client or a plugin like my HA-Commander which deals with data that changes very often. It can also be used for data that is slow to retrieve, presenting the user with partial data may feel snappier.

**Reply by pc223** [reply]
I still not entirely got it. 

So, for example, we make a download plugin, download Youtube video for example.

1. User input query to Flow
2. Flow send query to plugin
3. Plugin query Youtube and return back video list
4. User choose video need to be downloaded
5. Flow send download request to plugin
6. Plugin create a thread/process to handle download
7. ...What next?

**Reply by Garulf** [reply]
Rerun wouldn't work for a plugin like that. Its more for a plugin that shows your the status of your downloads in a torrent server for example. While the results are pulled up the user can see the download percentage and download rate update as they select their choice.

I've used this feature in alfred to update icons in the background. My plugin set a background task to download icons while alfred presented usable data to the user. Each time the launcher refreshed it updated with completed downloaded icons.

**Reply by Garulf** [reply]
I hope that makes sense

**Reply by taooceros** [reply]
C#plugin do have way to update results, but that's hard for python plugin to do.

**Reply by Garulf** [reply]
> C#plugin do have way to update results, but that's hard for python plugin to do.

No way for JsonRPC to schedule a rerun?

**Reply by taooceros** [reply]
Currently, no, but probably in the future.

**Reply by pc223** [reply]
> I hope that makes sense

So, kind of "live" result right? Something like this?

1. User input query progress download of a video
2. Flow send query to Plugin
3. Plugin return back current progress
4. Flow see "re-run" field and automatically re-run the query after x seconds
5. Back to step 2

**Reply by Garulf** [reply]
> > I hope that makes sense
> 
> So, kind of "live" result right? Something like this?
> 
> 1. User input query progress download of a video
> 2. Flow send query to Plugin
> 3. Plugin return back current progress
> 4. Flow see "re-run" field and automatically re-run the query after x seconds
> 5. Back to step 2

Exactly!
Once the plugin is complete it can simply ommit the rerun from results.

**Reply by taooceros** [reply]
So what about user change query? Do Flow still rerun the query?

**Reply by Garulf** [reply]
In that case the scheduled rerun would be canceled as its no longer needed

**Reply by taooceros** [reply]
Sounds great. I will write a implementation proposal later.

**Reply by Garulf** [reply]
Awesome! Thanks for everything!

**Reply by pc223** [reply]
This's pretty cool in terms of functionality, but the solution feel a bit hackish (quicker to implement right now but can be annoying later), can we find better solution? :smile:

The main idea is we need live result, result that can change content even while user is "choosing" (Ctrl+P/Ctrl+N)

A obivious design is make plugin a long running process,

After that we need a kind of "heartbeat", Flow will polling the plugin process for new content.

@taooceros I remember we have proposal to change the plugins to long running process right? (long running instead of the current one-shot style)

**Reply by taooceros** [reply]
Yes we do, but haven't had a complete api proposal for that.

**Reply by pc223** [reply]
Probably we should just try this re-run approach first, seems much quicker to implement, if waiting for a "proper" solution, I think it's gonna takes a long time 😄 early deliver early feedback stuff.

**Comment by taooceros** [comment]
## Proposal

- Let JsonRPCPlugin implement `IResultUpdate` (Or fire up ChangeQuery)
   - [ ] implement `IResultUpdate`
   - [ ] fire up ChangeQuery
- Response Format
```json
{
  "result": [],
  "debugMessage": "",
  "rerun": "an integer indicate delay(0 or -1 represent no rerun)"
}
```
- Rerun Request Format
```json
{
  "method": "query_rerun/query_update/query_refresh/...",
  "parameter": ["Query Search (should be the same as query)"]
}
```
- Detail Implementation specific
  - Include ID in request model indicates how much rerun has been run?
  - Use `async/await` to implement or use `Channel` to implement the task queue
  - Let JsonRPCPlugin (which is the base class) handles the rerun or let specific implementation to handle the rerun (e.g. PythonPlugin)

**Reply by taooceros** [reply]
@Flow-Launcher/team Thoughts?

**Reply by jjw24** [reply]
Instead of implementing for JSON-RPC plugins, how complex will it be to do it for all plugins via API?

**Reply by taooceros** [reply]
> Instead of implementing for JSON-RPC plugins, how complex will it be to do it for all plugins via API?

It is already applicable for C# plugin

**Reply by jjw24** [reply]
oh yeah lol 👍

**Reply by jjw24** [reply]
with the implementation, we will need somewhere to show what plugins have scheduled re-runs or currently re-running. we actually need to be very explicit about these re-runs because everything around flow has always been manually triggered with the exception of the automatic update, having the ability to do re-runs could open the door for plugins to do background stuff without the user's knowledge. So need to be very transparent and explicit about the re-runs, especially limiting the scope of the implementation.

**Reply by taooceros** [reply]
Currently the plan is rerun will stop when user change the query, so it is not a backdoor for background task, but just give jsonrpc plugin a way to update their results.

**Reply by pc223** [reply]
> Currently the plan is rerun will stop when user change the query, so it is not a backdoor for background task, but just give jsonrpc plugin a way to update their results.

I prefer this approach for right now, seems much simpler to implement than a long running background process. Although the long running process can be cool in the future.

**Reply by jjw24** [reply]
> rerun will stop when user change the query

oh ok, sounds good

**Comment by deefrawley** [comment]
I understand the concept here, but it seems like a major change for limited use cases. With Flow, it all happens with the query window - what task would users run there, and wait for updated results without letting the query window lose focus, where they wouldn't simply fire up a browser or dedicated app to handle this better?

**Reply by pc223** [reply]
I think it's not a major change, because it just re-run the current query after a interval.

And it can make the plugin more dynamic. For example the Currency plugin can have "live" result, like the currency conversion result "dancing" after every 1 sec, seems pretty cool.

Another example is the ProcessKiller plugin, with this feature, we can have CPU%, MEM%, IO rates, network rates,... "dancing" while users choosing process to kill 💀 :smile:

**Reply by deefrawley** [reply]
Ok, if it's not a huge change. As always I'll be keen to see what people do with it

**Reply by deefrawley** [reply]
I've actually thought of a way to use this. I played around with the psutils python library to do a quick system stat plugin (OS, disks, RAM and usage, CPU and usage) but to get the CPU % you need to make at least two calls to the function so it can measure between the two times. Plugin rerun achieves this and gives you live system stats so you could start a process, open Flow and keep an eye on things as it ran.

**Reply by taooceros** [reply]
> Another example is the ProcessKiller plugin, with this feature, we can have CPU%, MEM%, IO rates, network rates,... "dancing" while users choosing process to kill 💀 😄

Actually, C# plugin do support this feature currently. So we can implement this feature directly to it.

**Reply by Zeroto521** [reply]
> we can have CPU%, MEM%, IO rates, network rates,...

These things are status, long-lived application. Flow is a launcher that is used after called, short-lived application. 
This is not a good example.

![image](https://user-images.githubusercontent.com/25895405/128961086-c1641f19-b65d-4040-bfda-1ea1f396529a.png)

**Reply by jjw24** [reply]
i personally think it's fine,  there should be no constrain on the things that flow do should be launcher only. There are times i would like to see the battery percentage, cpu usage, memory etc in flow and having them on the status bar is one way but seeing them in the launcher is another way.

**Reply by Garulf** [reply]
A good example would be a process killer. The plugin could show memory/CPU hog's and with "live" information the user can better determine which process to kill.

**Comment by pc223** [comment]
Try out the current implementation of @taooceros here https://github.com/Flow-Launcher/Flow.Launcher/pull/648


<img width=350 src="https://user-images.githubusercontent.com/10551242/129316479-4605071d-7af3-46ba-9944-19adc203d10a.gif">
